sdk/python/kfp/compiler/_data_passing_rewriter.py

```diff
@@ -0,0 +1,357 @@
+    template_input_to_parent_task_outputs = {} # (task_template_name, task_input_name) -> Set[(upstream_template_name, upstream_output_name)]
+    template_input_to_parent_constant_arguments = {} #(task_template_name, task_input_name) -> Set[argument_value] # Unused
+    dag_output_to_parent_template_outputs = {} # (dag_template_name, output_name) -> Set[(upstream_template_name, upstream_output_name)]
+
+    for template in dag_templates:
+        dag_template_name = template['name']
+        # Indexing task arguments
+        dag_tasks = template['dag']['tasks']
+        task_name_to_task = {task['name']: task for task in dag_tasks}
+        task_name_to_template_name = {task['name']: task['template'] for task in dag_tasks}
+        for task in dag_tasks:
+            task_template_name = task['template']
+            parameter_arguments = task.get('arguments', {}).get('parameters', {})
+            for parameter_argument in parameter_arguments:
+                task_input_name = parameter_argument['name']
+                argument_value = parameter_argument['value']
+
+                argument_placeholder_parts = deconstruct_single_placeholder(argument_value)
+                if not argument_placeholder_parts: # Argument is considered to be constant string
+                    template_input_to_parent_constant_arguments.setdefault((task_template_name, task_input_name), set()).add(argument_value)
+
+                placeholder_type = argument_placeholder_parts[0]
+                if placeholder_type == 'inputs':
+                    assert argument_placeholder_parts[1] == 'parameters'
+                    dag_input_name = argument_placeholder_parts[2]
+                    template_input_to_parent_dag_inputs.setdefault((task_template_name, task_input_name), set()).add((dag_template_name, dag_input_name))
+                elif placeholder_type == 'tasks':
+                    upstream_task_name = argument_placeholder_parts[1]
+                    assert argument_placeholder_parts[2] == 'outputs'
+                    assert argument_placeholder_parts[3] == 'parameters'
+                    upstream_output_name = argument_placeholder_parts[4]
+                    upstream_template_name = task_name_to_template_name[upstream_task_name]
+                    template_input_to_parent_task_outputs.setdefault((task_template_name, task_input_name), set()).add((upstream_template_name, upstream_output_name))
+                elif placeholder_type == 'item' or placeholder_type == 'workflow' or placeholder_type == 'pod':
+                    # Treat loop variables as constant values
+                    # workflow.parameters.* placeholders are not supported, but the DSL compiler does not produce those.
+                    template_input_to_parent_constant_arguments.setdefault((task_template_name, task_input_name), set()).add(argument_value)
+                else:
+                    ...
```

Comment:
Maybe silly Q, does this line actually do anything?

Reply:
This is [Ellipsis](https://docs.python.org/3/library/constants.html#Ellipsis)
For some reason I thought it will be raising `NotImplementedError`, but that's not the case.
I'll replace it with `raise NotImplementedError`.

Reply:
Done.